File: till/src/main/java/com/wallee/android/till/sample/till/AuthorizeTransactionActivity.java
package com.wallee.android.till.sample.till;

import android.content.Context;
import android.os.Bundle;
import android.os.SystemClock;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.InputMethodManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.wallee.android.till.sample.till.common.GetAllCurrenciesUseCase;
import com.wallee.android.till.sample.till.databinding.AuthorizeTransactionActivityBinding;
import com.wallee.android.till.sample.till.model.Language;
import com.wallee.android.till.sample.till.model.Languages;
import com.wallee.android.till.sdk.ApiClient;
import com.wallee.android.till.sdk.TillLog;
import com.wallee.android.till.sdk.data.LineItem;
import com.wallee.android.till.sdk.data.Transaction;
import com.wallee.android.till.sdk.data.TransactionProcessingBehavior;
import java.math.BigDecimal;
import java.util.Currency;
import java.util.List;

public class AuthorizeTransactionActivity extends AppCompatActivity {

    private ApiClient client;
    private String languageCode;

    private AuthorizeTransactionActivityBinding binding;
    private long lastClickTime = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = AuthorizeTransactionActivityBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
        setUpLanguagesSpinner();
        setupCurrencySpinner();
        setupOnClickListener();
        client = new ApiClient(new MockResponseHandler(this));
        client.bind(this);
        TillLog.debug("SampleApp: Wallee ApiClient is just bound!");
    }

    private void authorizeTransaction() {
        String amountString = getTextAsString(binding.editTextAmount);
        String currencyString = binding.currencySpinner.getSelectedItem().toString();
        String customTextString = getTextAsString(binding.editTextCustomText);
        String transactionRefNumber = getTextAsString(binding.editTextTransactionRef);

        // For credit transactions the value has to be negative Ex: -10.00

        if(binding.shouldReserve.isChecked() && binding.shouldAdjustReservation.isChecked()) {
            Toast.makeText(this, "Choose one: Reservation or Reservation adjustment", Toast.LENGTH_LONG).show();
        } else if (amountString.isEmpty()) {
            Toast.makeText(this, "Amount field is empty", Toast.LENGTH_LONG).show();
        } else if (binding.shouldAdjustReservation.isChecked() && transactionRefNumber.isEmpty()){
            Toast.makeText(this, "Reserve reference field is empty", Toast.LENGTH_LONG).show();
        } else {
            hideKeyboardFrom(AuthorizeTransactionActivity.this);

            TransactionProcessingBehavior behavior;
            if(binding.shouldReserve.isChecked()){
                behavior = TransactionProcessingBehavior.RESERVE;
            } else if(binding.shouldAdjustReservation.isChecked()){
                behavior = TransactionProcessingBehavior.ADJUST_RESERVATION;
            } else {
                behavior = TransactionProcessingBehavior.COMPLETE_IMMEDIATELY;
            }

            List<LineItem> lineItems = new LineItem.ListBuilder("foo", new BigDecimal(amountString))
                    .getCurrent()
                    .setName("bar")
                    .getListBuilder()
                    .build();

            Transaction transaction;
            Transaction.Builder transactionBuilder = new Transaction.Builder(lineItems)
                    .setCurrency(Currency.getInstance(currencyString))
                    .setInvoiceReference("1")
                    .setMerchantReference("MREF-123")
                    .setTransactionProcessingBehavior(behavior)
                    .setGeneratePanToken(binding.generatePanToken.isChecked());
            if (!customTextString.isEmpty()) {
                transactionBuilder.setCustomText(customTextString);
            }
            if (binding.selectLanguage.isChecked()) {
                transactionBuilder.setLanguage(languageCode);
            }
            if(binding.shouldAdjustReservation.isChecked()) {
                transactionBuilder.setTransactionRefNumber(transactionRefNumber);
            }
            transaction = transactionBuilder.build();
            TillLog.debug("VSD Start Transaction of amount  -> " + amountString);
            try {
                client.authorizeTransaction(transaction);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private void setUpLanguagesSpinner() {
        List<Language> languages = new Languages().getLanguages();
        ArrayAdapter arrayAdapter
                = new ArrayAdapter(
                this,
                android.R.layout.simple_spinner_item,
                languages);
        arrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_item);
        binding.languageSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                languageCode = languages.get(position).getCode();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
            }
        });
        binding.languageSpinner.setAdapter(arrayAdapter);
    }

    private void setupCurrencySpinner() {
        GetAllCurrenciesUseCase getAllCurrenciesUseCase = new GetAllCurrenciesUseCase();
        ArrayAdapter<com.wallee.android.till.sample.till.model.Currency> arrayAdapter = new ArrayAdapter<>(this,
                R.layout.item_simple, getAllCurrenciesUseCase.execute());
             binding.currencySpinner.setAdapter(arrayAdapter);
    }

    private void setupOnClickListener() {
        binding.selectLanguage.setOnCheckedChangeListener((buttonView, isChecked) -> setVisibleOrGone(binding.languageSpinner, isChecked));

        binding.authorizeButton.setOnClickListener(v -> {
            if (SystemClock.elapsedRealtime() - lastClickTime < 1000) {
                return;
            }
            lastClickTime = SystemClock.elapsedRealtime();
            authorizeTransaction();
        });

        binding.returnButton.setOnClickListener(v -> finish());

        binding.authorizeTransactionParent.setOnClickListener(v -> hideKeyboardFrom(AuthorizeTransactionActivity.this));
        binding.shouldAdjustReservation.setOnCheckedChangeListener((buttonView, isChecked) -> setVisibleOrGone(binding.editTextTransactionRef, isChecked));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
        if (client != null) {
            client.unbind(this);
        }
    }

    private String getTextAsString(android.widget.EditText editText) {
        return editText.getText().toString().trim();
    }

    private void setVisibleOrGone(View view, boolean isVisible) {
        view.setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

    private void hideKeyboardFrom(Context context) {
        InputMethodManager inputMethodManager = (InputMethodManager) context.getSystemService(Context.INPUT_METHOD_SERVICE);
        if (inputMethodManager != null) {
            inputMethodManager.hideSoftInputFromWindow(binding.authorizeTransactionParent.getWindowToken(), 0);
        }
    }
}

File: till/src/main/java/com/wallee/android/till/sample/till/AuthorizeTransactionActivity.kt
package com.wallee.android.till.sample.till

import android.os.Bundle
import android.os.SystemClock
import android.util.Log
import android.view.View
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.CompoundButton
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.wallee.android.till.sample.till.common.GetAllCurrenciesUseCase
import com.wallee.android.till.sample.till.common.getTextAsString
import com.wallee.android.till.sample.till.common.setVisibleOrGone
import com.wallee.android.till.sample.till.databinding.AuthorizeTransactionActivityBinding
import com.wallee.android.till.sample.till.model.Languages
import com.wallee.android.till.sdk.ApiClient
import com.wallee.android.till.sdk.TillLog
import com.wallee.android.till.sdk.data.LineItem
import com.wallee.android.till.sdk.data.Transaction
import com.wallee.android.till.sdk.data.TransactionProcessingBehavior
import java.math.BigDecimal
import java.util.*

class AuthorizeTransactionActivity : AppCompatActivity() {

    private var client: ApiClient? = null
    private var languageCode: String? = null

    private var _binding: AuthorizeTransactionActivityBinding? = null

    // This property is only valid between onCreateView and onDestroyView.
    private val binding get() = _binding!!
    private var lastClickTime: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        _binding = AuthorizeTransactionActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
        setUpLanguagesSpinner()
        setupCurrencySpinner()
        setupOnClickListener()
        client = ApiClient(MockResponseHandler(this))
        client?.bind(this)
        TillLog.debug("SampleApp: Wallee ApiClient is just bound!")
    }

    private fun authorizeTransaction() {
        with(binding) {
            val amountString = editTextAmount.getTextAsString()
            val currencyString = currencySpinner.selectedItem.toString()
            val customTextString = editTextCustomText.getTextAsString()

            if (amountString.isBlank()) {
                Toast.makeText(
                    this@AuthorizeTransactionActivity,
                    "Amount field is empty",
                    Toast.LENGTH_LONG
                ).show()
            } else {
                val lineItems = LineItem.ListBuilder("foo", BigDecimal(amountString))
                    .current
                    .setName("bar")
                    .listBuilder
                    .build()

                val transaction: Transaction
                val transactionBuilder = Transaction.Builder(lineItems)
                    .setCurrency(Currency.getInstance(currencyString))
                    .setInvoiceReference("1")
                    .setMerchantReference("MREF-123")
                    .setTransactionProcessingBehavior(if (shouldReserve.isChecked) TransactionProcessingBehavior.RESERVE else TransactionProcessingBehavior.COMPLETE_IMMEDIATELY)
                    .setGeneratePanToken(generatePanToken.isChecked)
                if (customTextString.isNotEmpty()) {
                    transactionBuilder.customText = customTextString
                }
                if (selectLanguage.isChecked) transactionBuilder.language = languageCode
                transaction = transactionBuilder.build()
                TillLog.debug("VSD Start Transaction of amount  -> $amountString")
                try {
                    client?.authorizeTransaction(transaction)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun setUpLanguagesSpinner() {
        val languages = Languages().languages
        val arrayAdapter = ArrayAdapter(
            this,
            android.R.layout.simple_spinner_item,
            languages
        )
        arrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_item)
        binding.languageSpinner.onItemSelectedListener = object : OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View,
                position: Int,
                id: Long
            ) {
                languageCode = languages[position].code
            }

            override fun onNothingSelected(p0: AdapterView<*>?) {}
        }
        binding.languageSpinner.adapter = arrayAdapter
    }

    private fun setupCurrencySpinner() {
        val getAllCurrenciesUseCase = GetAllCurrenciesUseCase()
        val currencies = getAllCurrenciesUseCase.execute()
        val arrayAdapter = ArrayAdapter(
            this,
            R.layout.item_simple,
            currencies
        )
        binding.currencySpinner.adapter = arrayAdapter
    }

    private fun setupOnClickListener() {
        with(binding) {
            selectLanguage.setOnCheckedChangeListener { _: CompoundButton?, isChecked: Boolean ->
                languageSpinner.setVisibleOrGone(isChecked)
            }
            authorizeButton.setOnClickListener {
                // preventing double, using threshold of 1000 ms
                if (SystemClock.elapsedRealtime() - lastClickTime < 1000){
                    return@setOnClickListener
                }
                lastClickTime = SystemClock.elapsedRealtime();
                authorizeTransaction()
            }
            returnButton.setOnClickListener { finish() }
            authorizeTransactionParent.setOnClickListener {
                Utils.hideKeyboardFrom(this@AuthorizeTransactionActivity)
            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        _binding = null
        client?.unbind(this)
    }
}
